java-manta-client/src/main/java/com/joyent/manta/client/multipart/EncryptionStateRecorder.java

```diff
@@ -0,0 +1,176 @@
+import com.joyent.manta.exception.MantaReflectionException;
+import com.joyent.manta.util.CipherCloner;
+import com.joyent.manta.util.Cloner;
+import com.joyent.manta.util.HmacCloner;
+import com.joyent.manta.util.HmacOutputStream;
+import org.apache.commons.lang3.reflect.FieldUtils;
+import org.bouncycastle.crypto.macs.HMac;
+import org.bouncycastle.jcajce.io.CipherOutputStream;
+
+import javax.crypto.Cipher;
+import java.io.OutputStream;
+import java.lang.reflect.Field;
+
+/**
+ * Helper class for recording state of encryption and authentication state during in-progress MPU uploads.
+ * Used to enable retry of {@link EncryptedMultipartManager#uploadPart} in case of network failure.
+ * Like the <a href="https://sourcemaking.com/design_patterns/memento">Memento Pattern</a> but with less respect
+ * for encapsulation.
+ */
+public class EncryptionStateRecorder {
+
+    /**
+     * {@link EncryptionState} to make rewindable.
+     */
+    private final EncryptionState encryptionState;
+
+    /**
+     * Whether or not separate authentication is in use.
+     * True indicates we need to back up HMAC state.
+     * False indicates the Cipher has built-in authentication.
+     */
+    private final boolean usesHmac;
+
+    /**
+     * Stashed HMac state we can restore to.
+     */
+    private HMac savedHmac = null;
+
+    /**
+     * Stashed Cipher state we can restore to.
+     */
+    private Cipher savedCipher = null;
+
+    /**
+     * Reference to {@link HmacOutputStream}'s {@link HMac} field.
+     */
+    private static final Field FIELD_HMACOUTPUTSTREAM_HMAC =
+            FieldUtils.getField(HmacOutputStream.class, "hmac", true);
+
+    /**
+     * Reference to {@link HmacOutputStream}'s {@link OutputStream} field.
+     */
+    private static final Field FIELD_HMACOUTPUTSTREAM_OUT =
+            FieldUtils.getField(HmacOutputStream.class, "out", true);
+
+    /**
+     * Reference to {@link EncryptionContext}'s {@link Cipher} field.
+     */
+    private static final Field FIELD_ENCRYPTIONCONTEXT_CIPHER =
+            FieldUtils.getField(EncryptionContext.class, "cipher", true);
+
+    /**
+     * Reference to {@link CipherOutputStream}'s {@link Cipher} field.
+     */
+    private static final Field FIELD_CIPHEROUTPUTSTREAM_CIPHER =
+            FieldUtils.getField(CipherOutputStream.class, "cipher", true);
+
+    /**
+     * {@link HMac} cloning helper object.
+     */
+    private static final Cloner<HMac> CLONER_HMAC = new HmacCloner();
+
+    /**
+     * {@link Cipher} cloning helper object.
+     */
+    private static final Cloner<Cipher> CLONER_CIPHER = new CipherCloner();
+
+    /**
+     * Construct an EncryptionStateRecorder which can be used to rewind the state of the given {@link EncryptionState}.
+     *
+     * @param encryptionState object to extract state from (and potentially write back to)
+     */
+    public EncryptionStateRecorder(final EncryptionState encryptionState) {
```

Comment:
The class design here seems oddly inverted. I would expect a single `EncryptionStateRecorder` instance that I would leave laying around. It would then take a `EncryptionState` parameter on `record()` and then return a state object. Then `rewind()` would parse that state object and give me a new `EncryptionState` object.

Help convince me as the reader that your mode of operation is better and why in the comments and code.